### ISSUE TITLE
Discussing MultipartUploader's handling of seekable streams

### ISSUE THREAD

**Issue by umesecke** [aws/aws-sdk-php#1101]
The multipart uploader does some 'special' handling of streams I'd like to discuss.

Streams are only treated as seekable if the underlying resource is a filehandle ("wrapper_type = plainfile", see [AbstractUploader::getUploadCommands()](https://github.com/aws/aws-sdk-php/blob/master/src/Multipart/AbstractUploader.php#L46)). In the MultipartUploader implementations for S3 and Glacier there is even access to the underlying url and a new stream is created opening the file for each part, bypassing the original stream (see [MultipartUploader::createPart()](https://github.com/aws/aws-sdk-php/blob/master/src/S3/MultipartUploader.php#L90)).

As a result, all streams that don't have an underlying file are first copied into a new stream backed by "php://temp". That means they are written into a temporary file in a location provided by sys_get_temp_dir(). PHP writes the data to disk if the size is larger than 2 MB which is the case for every but maybe the last part of a multipart upload.

This all brings (a) latency because of copying data to disk, (b) additional space requirements on disk, (c) additional open filehandles and (d) the risk of the disk running full. This is done even if the stream is declared as seekable by it's implementation. And in case of an error and restarting the uploader all this buffering is repeated for parts that are still to be transmitted.

Is there a reason for doing this and depending on the underlying source of a stream? Shouldn't it be the responsibility of the stream's implementation to handle all that?

The only reason I can think of is that multiple streams all have their own seek position and don't interfere with other consumers of the underlying file in an async program execution context. But shouldn't it be sufficient to call seek() before reading from the stream to make sure it is back in the right position? I am not aware of current multithreading features in other php implementations (HHVM, etc) but at least in plain old PHP I have no situation in mind where a seek() directly followed by a read() in the same "unit of work" should be disturbed by other code re-seeking the stream.

For non-seekable streams I can see that this buffering is necessary because the parts are uploaded concurrently. But if the source stream is seekable I don't think all this is needed.

In my case (and the reason I am writing all this) I have a seekable stream implementation that handles reading parts of a large file from multiple third-party API calls. Providing that stream to the multipart uploader first buffers the parts on disk which has the high risk of the disk running full in case of multiple concurrent requests in my application. This leaves me with either implementing multipart uploads using the low-level API myself or creating a stream wrapper for my stream implementation faking to be a "plainfile".

**Comment by cjyclaire** [comment]
@umesecke Thank you for the feedback and thoughts! I see your case, so firstly, as you have mentioned, this buffering is necessary for non-seekable streams definitely. So the main concern is why SDK checks `'plainfile'` type in qualification for a seekable stream.

So, `MultipartUpload` needs to ensure each part matches precisely with stream resource or there could be discrepancy between contents. SDK uses [`LimitStream`](https://github.com/guzzle/psr7/blob/master/src/LimitStream.php) in Guzzle to achieve that. For php [`fread`](http://php.net/manual/en/function.fread.php), there could be unexpected behavior for streams not represents a plain file:
>if the stream is read buffered and it does not represent a plain file, at most one read of up to a number of bytes equal to the chunk size (usually 8192) is made; depending on the previously buffered data, the size of the returned data may be larger than the chunk size.

**Comment by umesecke** [comment]
@cjyclaire Ah, thanks for the hint on `fread`'s behaviour! I wasn't aware of this.

So it makes sense to treat all non-plainfile streams as non-seekable and reading them from start to end because neither does `Psr7\LimitStream::read()` enforce the `$length` parameter nor does `StreamInterface::read()` says anything about that.

But of course we could easily create such an implementation of a limiting stream that enforces the limit in `read()`. Worst case scenario the underlying stream has to fetch up to one chunk of data again. If the implementation is seekable it should be able to handle this.

I can understand that there may be performance implications for some implementations if reading is not exactly aligned to chunk size. But in my case I would gladly trade those off for not having to buffer gigabytes of data to disk.

Perhaps we could add an option that enables those extra limits and in return rely fully on the stream's seeking ability without buffering to disk. Maybe wrapping the stream into a stream decorator that acts as a marker for me telling the sdk that I have understood the possible implications and that I am able to handle that. ;)

**Comment by cjyclaire** [comment]
@umesecke Appreciate your feedback, your case does have a point worth considering, for limit bytes read, there is a PR guzzle/psr7#105 opened with implementation of a `ByteCountingStream` that may help with the situation. I'll tag this as a feature request and track it in our backlog.

**Comment by umesecke** [comment]
@cjyclaire sounds great! thanks

**Comment by mtdowling** [comment]
Would it be possible to buffer in memory by changing the memory to disk cutoff in the temp stream? Perhaps we should just update the amount of data cached in memory to 5 MB (or even allow it to be configurable).

http://php.net/manual/en/wrappers.php.php

> The memory limit of php://temp can be controlled by appending /maxmemory:NN, where NN is the maximum amount of data to keep in memory before using a temporary file, in bytes.

**Comment by cjyclaire** [comment]
@mtdowling That's a good point, make sense to me. I'll dig deeper into this. Allowing the cache size is indeed a good point worth considering.

**Comment by umesecke** [comment]
I also did some further investigation and realized that the buffering is not done for the whole stream upfront but only just for the parts that are going to be uploaded next (I had to catch up with generator functions in PHP). So most of my points from the beginning seem to be obsolete. As the temporary file should be closed (and deleted) by GC after the upload of a part is finished the impact on disk size should be minimal (partsize x concurrent uploads). In that case the benefits of preloading the parts IMHO outweight the downsides by far.

So for my use-case the only thing left is the case of resuming a failed upload. As the stream is not handled as seekable, the data of the already successfully uploaded parts is read again from the stream to advance the offset to the beginning of the part to be re-uploaded (see getUploadCommands() in AbstractUploader).

I can see the possible problems when the amount of data puffered for a part does not equal the part size (`fread` reading up to a chunksize of data) vs. skipping the part by the exact part size. Maybe we can track the offsets of the already uploaded parts in the upload state so that even if the offsets are somewhat skewed we are still able to seek to the beginning of the next part.
